spec/lib/requests_rate_limiter_spec.rb

```diff
@@ -0,0 +1,280 @@
+        ins.apply_limits! do
+          ins.request.env['DISCOURSE_IS_ASSET_PATH'] = false
+        end
+      }.to change { ins.limiter_10_secs.remaining }.by(-1)
+        .and change { ins.limiter_60_secs.remaining }.by(-1)
+        .and change { ins.assets_limiter_10_secs.remaining }.by(0)
+    end
+
+    it "rolls back normal rate limiters if the request is an assets request" do
+      ins = instance
+      expect {
+        ins.apply_limits! do
+          ins.request.env['DISCOURSE_IS_ASSET_PATH'] = true
+        end
+      }.to change { ins.limiter_10_secs.remaining }.by(0)
+        .and change { ins.limiter_60_secs.remaining }.by(0)
+        .and change { ins.assets_limiter_10_secs.remaining }.by(-1)
+    end
+
+    it "does not yield if a limit is reached and rate limit mode is block" do
+      global_setting :max_reqs_per_ip_per_10_seconds, 1
+      x = 0
+      ins = instance
+
+      # 2nd iteration is rate limited
+      2.times do
+        ins.apply_limits! do
+          x += 1
+        end
+      end
+      expect(x).to eq(1)
+      expect(Rails.logger.warnings).to eq(0)
+    end
+
+    it "yields if rate limits are skipped" do
+      global_setting :max_reqs_per_ip_mode, "none"
+      x = 0
+      instance.apply_limits! do
+        x += 1
+      end
+      expect(x).to eq(1)
+    end
+
+    it "yields if a rate limit is reached and rate limit mode is warn" do
+      global_setting :max_reqs_per_ip_mode, "warn"
+      global_setting :max_reqs_per_ip_per_10_seconds, 1
+      ins = instance
+      x = 0
+      2.times do
+        ins.apply_limits! do
+          x += 1
+        end
+      end
+      expect(x).to eq(2)
+      expect(Rails.logger.warnings).to eq(1)
+    end
+
+    it "does not yield if a rate limit is reached and rate limit mode is warn+block" do
+      global_setting :max_reqs_per_ip_mode, "warn+block"
+      global_setting :max_reqs_per_ip_per_10_seconds, 1
+      ins = instance
+      x = 0
+      2.times do
+        ins.apply_limits! do
+          x += 1
+        end
+      end
+      expect(x).to eq(1)
+      expect(Rails.logger.warnings).to eq(1)
+    end
+
+    it "returns yield results if no rate limit is reached or rate limits are disabled" do
+      res = instance.apply_limits! do
+        "hello world"
+      end
+      expect(res).to eq("hello world")
+
+      global_setting :max_reqs_per_ip_mode, "none"
+      res = instance.apply_limits! do
+        "hello world2"
+      end
+      expect(res).to eq("hello world2")
+    end
+
+    it "returns the right error response when applying per-user per-10-seconds " \
+    "rate limits" do
+      global_setting :max_reqs_per_ip_per_10_seconds, 1
+      global_setting :skip_per_ip_rate_limit_trust_level, 2
+      ins = instance({}, user)
+      x = 0
+      ins.apply_limits! do
+        x += 1
+      end
+      response = ins.apply_limits! do
+        x += 1
+      end
+      expect(x).to eq(1)
+      assert_rate_limit_response(response, "id_10_secs_limit")
```

Comment:
I'm confused again here. The user has a trust level of 2 so that rate limit should apply to the user?

Reply:
I think this question is the same as https://github.com/discourse/discourse/pull/14706#discussion_r744426682?

Reply:
Yup :+1: I'll resovle this comment